test/integration/helpers.bash

```diff
@@ -86,32 +91,55 @@ function swarm_manage() {
 		discovery="$@"
 	fi
 
-	$SWARM_BINARY manage -H $SWARM_HOST $discovery &
+	"$SWARM_BINARY" manage -H "$SWARM_HOST" --heartbeat=1 "$discovery" &
 	SWARM_PID=$!
-	wait_until_reachable $SWARM_HOST
+	wait_until_reachable "$SWARM_HOST"
+	retry 10 1 check_swarm_nodes
 }
 
-# Start swarm join for every engine with the discovery as parameter
+# swarm join every engine created with `start_docker`.
+#
+# It will wait until all nodes are visible in discovery (`swarm list`) before
+# returning and will fail if that's not the case after a certain time.
+#
+# It can be called multiple times and will only join new engines started with
+# `start_docker` since the last `swarm_join` call.
 function swarm_join() {
-	local i=0
-	for h in ${HOSTS[@]}; do
-		echo "Swarm join #${i}: $h $@"
-		$SWARM_BINARY join --addr=$h "$@" &
+	local current=${#SWARM_JOIN_PID[@]}
+	local nodes=${#HOSTS[@]}
+	local addr="$1"
+	shift
+
+	# Start the engines.
+	local i
+	echo "current: $current | nodes: $nodes" > log
+	for ((i=current; i < nodes; i++)); do
+		local h="${HOSTS[$i]}"
+		echo "Swarm join #${i}: $h $addr"
+		"$SWARM_BINARY" join --addr="$h" "$addr" &
 		SWARM_JOIN_PID[$i]=$!
-		((++i))
 	done
-	retry 30 1 [ -n "$(docker_swarm info | grep -q 'Nodes: $i')" ]
+	retry 10 0.5 check_discovery_nodes "$addr"
```

Comment:
This won't fail on the CI but will fail on my machines in case of `token` as mine are quite far and network lagging will be occurred. Is there anyway to workaround this number? Perhaps make it configurable like `$SWARM_ROOT` ?